### DISCUSSION TITLE
Using GPU to enhance the speed

### DISCUSSION THREAD

**Discussion by JamesWeiChen** [lnccbrown/HSSM#397] [Ideas]
@frankmj Since HSSM is built on PyMC and incorporates features from the Python Bayesian ecosystem, it suggests it can utilize GPU acceleration indirectly through PyMC if appropriately configured. However, I think we cannot currently use GPU to boost the speed, right? Is it possible to have that feature in the future? Thank you!

Best,
James

**Comment by frankmj** [comment]
Yes you can certainly use GPUs and this should speed up processing
considerably especially when using LANs. If using Colab you can simply
switch the runtime to use GPU.

Michael

**Reply by JamesWeiChen** [reply]
@frankmj Thank you for your suggestion! Could you provide specific instructions on configuring and enabling GPU acceleration on a personal computer? If there are specific commands or steps, could you guide me through them? Thank you!

**Comment by frankmj** [comment]
I will let @Paul Xu ***@***.***>  respond on this issue :^)

**Comment by digicosmos86** [comment]
Hi @JamesWeiChen,

It's actually quite easy to use GPU acceleration with HSSM. All you need to do is to install JAX with GPU support. Detailed installation guide can be found [here](https://jax.readthedocs.io/en/latest/installation.html#pip-installation-nvidia-gpu-cuda-installed-via-pip-easier). Once you have installed JAX, proceed to install pymc and HSSM following our instruction [here](https://github.com/lnccbrown/HSSM?tab=readme-ov-file#installation).

Once everything is set up, JAX will automatically perform computation with GPU unless otherwise instructed. You just need to use a JAX-based sampler (e.g. set `sampler="nuts_numpyro"` when calling `model.sample()`).

If you are using colab with GPU, JAX with GPU support is installed by default. You can just install HSSM via `pip`. Please check out [this issue](https://github.com/lnccbrown/HSSM/issues/396) to get started.

Please let me know if you have any questions :)

Paul
